File: data_structures/hash_table/hash_table.go
package datastructures

type entry struct {
	key   string
	value int
	next  *entry
}

type HashTable struct {
	buckets []*entry
	size    int
	cap     int
}

func NewHashTable(capacity int) *HashTable {
	return &HashTable{
		buckets: make([]*entry, capacity),
		size:    0,
		cap:     capacity,
	}
}

func (h *HashTable) hash(key string) int {
	sum := 0
	for i := 0; i < len(key); i++ {
		sum += int(key[i])
	}
	return sum % h.cap
}

func (h *HashTable) Put(key string, value int) {
	index := h.hash(key)
	curr := h.buckets[index]

	for curr != nil {
		if curr.key == key {
			curr.value = value
			return
		}
		curr = curr.next
	}

	newEntry := &entry{
		key:   key,
		value: value,
		next:  h.buckets[index],
	}
	h.buckets[index] = newEntry
	h.size++
}

func (h *HashTable) Get(key string) (int, bool) {
	index := h.hash(key)
	curr := h.buckets[index]

	for curr != nil {
		if curr.key == key {
			return curr.value, true
		}
		curr = curr.next
	}

	return 0, false
}

func (h *HashTable) Delete(key string) bool {
	index := h.hash(key)
	curr := h.buckets[index]

	if curr == nil {
		return false
	}

	if curr.key == key {
		h.buckets[index] = curr.next
		h.size--
		return true
	}

	for curr.next != nil {
		if curr.next.key == key {
			curr.next = curr.next.next
			h.size--
			return true
		}
		curr = curr.next
	}

	return false
}

func (h *HashTable) Contains(key string) bool {
	_, found := h.Get(key)
	return found
}

func (h *HashTable) Size() int {
	return h.size
}

func (h *HashTable) Keys() []string {
	keys := make([]string, 0, h.size)
	for _, bucket := range h.buckets {
		curr := bucket
		for curr != nil {
			keys = append(keys, curr.key)
			curr = curr.next
		}
	}
	return keys
}

File: data_structures/hash_table/python/hash_table.py
from __future__ import annotations


class _Entry:
    def __init__(self, key: str, value: int, next: _Entry | None = None):
        self.key = key
        self.value = value
        self.next = next


class HashTable:
    def __init__(self, capacity: int = 16):
        self._buckets: list[_Entry | None] = [None] * capacity
        self._size = 0
        self._cap = capacity

    def _hash(self, key: str) -> int:
        return sum(ord(c) for c in key) % self._cap

    def put(self, key: str, value: int) -> None:
        index = self._hash(key)
        curr = self._buckets[index]
        while curr is not None:
            if curr.key == key:
                curr.value = value
                return
            curr = curr.next
        self._buckets[index] = _Entry(key, value, self._buckets[index])
        self._size += 1

    def get(self, key: str) -> tuple[int, bool]:
        index = self._hash(key)
        curr = self._buckets[index]
        while curr is not None:
            if curr.key == key:
                return curr.value, True
            curr = curr.next
        return 0, False

    def delete(self, key: str) -> bool:
        index = self._hash(key)
        curr = self._buckets[index]
        if curr is None:
            return False
        if curr.key == key:
            self._buckets[index] = curr.next
            self._size -= 1
            return True
        while curr.next is not None:
            if curr.next.key == key:
                curr.next = curr.next.next
                self._size -= 1
                return True
            curr = curr.next
        return False

    def contains(self, key: str) -> bool:
        _, found = self.get(key)
        return found

    def size(self) -> int:
        return self._size

    def keys(self) -> list[str]:
        result: list[str] = []
        for bucket in self._buckets:
            curr = bucket
            while curr is not None:
                result.append(curr.key)
                curr = curr.next
        return result
